### PR TITLE
New module account_permanent_lock_move

### PR DESCRIPTION
**Pull request by mdietrichc2c**
New module, which adds a new lock date in Invoicing Settings.
This date is a more "permanent" lock, as it cannot be unset nor set back in the past ; once it is set, the moves prior to and including at this date cannot be created, modified, deleted in any way.

### CODE REVIEW

File: account_permanent_lock_move/models/res_config.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class AccountConfigSettings(models.TransientModel):
+    _inherit = 'account.config.settings'
+
+    permanent_lock_date = fields.Date(
+        string="Permanent Lock Date",
+        related='company_id.permanent_lock_date',
+        help="Non-revertible closing of accounts prior to and inclusive of "
+        "this date. Use it for fiscal year locking instead of ""Lock Date"".")
```
Comment (guewen) [inline account_permanent_lock_move/models/res_config.py]: `""` will have no effect (Python will considerate it as 2 strings to concatenate). You should use `\"` or surround the whole string with `'`.

File: account_permanent_lock_move/security/ir.model.access.csv [hunk 2]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_account_permanent_lock_date_wizard,account.permanent.lock.move,model_permanent_lock_date_wizard,account.group_account_manager,1,1,1,1
```
Comment (guewen) [inline account_permanent_lock_move/security/ir.model.access.csv]: TransientModels should not have access rights.

File: account_permanent_lock_move/wizard/permanent_lock_date_wizard.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                                 string='Company')
+
+    @api.multi
+    def confirm_date(self):
+        self.ensure_one()
+        company = self.company_id
+        if (company.permanent_lock_date and
+                self.lock_date < company.permanent_lock_date):
+            raise UserError(
+                _("You cannot set the permanent lock date in the past!")
```
Comment (guewen) [inline account_permanent_lock_move/wizard/permanent_lock_date_wizard.py]: Exclamation marks should be avoided in error messages (https://cybertext.wordpress.com/2011/11/18/error-messages-that-make-you-feel-stupid/)

File: account_permanent_lock_move/wizard/permanent_lock_date_wizard.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            )
+        # Then check if unposted moves are present before the date
+        moves = self.env['account.move'].search(
+            [('company_id', '=', self.id),
+                ('date', '<=', self.lock_date),
+                ('state', '=', 'draft')])
+        if moves:
+            raise UserError(
+                _("You cannot permanently lock since entries "
+                  "are still unposted before this date!")
```
Comment (guewen) [inline account_permanent_lock_move/wizard/permanent_lock_date_wizard.py]: Exclamation marks should be avoided in error messages (https://cybertext.wordpress.com/2011/11/18/error-messages-that-make-you-feel-stupid/)

**Comment by fclementic2c** [comment]
tests done:
- As non-adviser : I cannot post an entry before the lock date for non-adviser : Ok
- As non-adviser : I can post an entry after the lock date for non-adviser : Ok
- As non-adviser : I can reconcile a payment posted after the lock date for non-adviser (invoice prior to this date) : Ok
- As adviser : I can post an entry before the lock date for non-adviser : Ok
- As adviser : I can post an entry after the lock date for non-adviser : Ok
- As adviser : I can reconcile a payment posted after the lock date for non-adviser (invoice prior to this date) : Ok
- As adviser : I can post an entry before the lock date for adviser : Ok but last sentence of the message should be modified (since I am already the adviser):  
  'You cannot add/modify entries prior to and inclusive of the lock date 2016-05-01. **Check the company settings or ask someone with the 'Adviser' role**' 
- As adviser : I can post an entry after the lock date for adviser : Ok
- As adviser : I can reconcile a payment posted after the lock date for adviser (invoice prior to this date) : Ok
- 2 minor ergonomy points : 
- Please create a section in the Settings called 'Fiscal Year closing' to clarify for end-user : 
- Improve button Permanent locking please
  image

**Comment by guewen** [comment]
:+1: (code review, no test)

**Comment by fclementic2c** [comment]
:+1: 

Thank you

File: account_permanent_lock_move/README.rst [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+===================
+Permanent Lock Move
+===================
+
+This module extends the functionality of the basic "lock date"
+functionality in invoicing, in order to add a more definitive
+lock date (which does not allow to be set at a previous date,
```
Comment (elicoidal) [inline account_permanent_lock_move/README.rst]: which cannot be empty or set in the past, and only accept unposted moves)

File: account_permanent_lock_move/README.rst [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+or emptied, and only accepts unposted moves)
+
+Configuration
+=============
+
+To configure this module, you need to:
+
+* Go to Invoicing -> Settings, and set the Permanent Lock Date.
```
Comment (elicoidal) [inline account_permanent_lock_move/README.rst]: #. Go to Invoicing -> Settings
#. set the Permanent Lock Date.

File: account_permanent_lock_move/wizard/permanent_lock_date_wizard.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                _("You cannot set the permanent lock date in the past.")
+            )
+        # Then check if unposted moves are present before the date
+        moves = self.env['account.move'].search(
+            [('company_id', '=', self.id),
+                ('date', '<=', self.lock_date),
+                ('state', '=', 'draft')])
+        if moves:
+            raise UserError(
+                _("You cannot permanently lock since entries "
```
Comment (elicoidal) [inline account_permanent_lock_move/wizard/permanent_lock_date_wizard.py]: You cannot set the permanent lock date since...

**Comment by elicoidal** [comment]
some minor points.
:+1: 

Thanks for the module!

**Comment by mdietrichc2c** [comment]
@elicoidal thanks for the review ; I changed the phrasing accordingly.

**Comment by elicoidal** [comment]
@mdietrichc2c thanks for the contribution!

**Comment by sbidoul** [comment]
LGTM. Do you plan to add a test?

**Comment by mdietrichc2c** [comment]
@sbidoul I did not plan it, but doing one allowed me to spot a few cases.

**Comment by sbidoul** [comment]
> @sbidoul I did not plan it, but doing one allowed me to spot a few cases.

:smile: I love when writing a test reveals bugs!

Is the runbot failure related?

**Comment by mdietrichc2c** [comment]
Wrong decorator on `create` : it should be fixed now.

**Comment by sbidoul** [comment]
:+1: Thanks a lot!

**Comment by leemannd** [comment]
:+1:

**Comment by sbidoul** [comment]
Should `account_move_locking` be deprecated somehow?

**Pull request merged**